book/testing_code.md

```diff
@@ -321,12 +321,65 @@ Integration tests that check very specific outcomes will need to be updated with
 User acceptance tests are those that check that a high level user requirement has been met.
 In analysis, these are likely part of an end to end test that checks that the output is fit for purpose.
 
-```{todo}
-Discuss testing interface with external systems (e.g. database).
-Test that your code works, given the format of response that the system can give.
-Mocks?
+## Isolate code tests from external systems
+
+Testing code that interacts with an external system can be particularly challenging when you can't guarantee
+that the system will provide you with the same response each time; this could include code querying a database or
+making API requests, for example.
+
+Best practice is to seperate external system dependencies from your code tests as much as possible. This can mitigate against various risks, depending on your application:
+
+- Testing database interaction with a production database could result in damage to, or loss of, data.
+- Making real API calls when testing a function that handles requests could incur unintended monetary costs.
```

Comment:
These are 2 of the most common reasons to mock. There is another, less discussed and important reason to mock: To enforce isolation in unit tests. To enforce loose coupling in your own source code, thorough mocking would mock dependencies in your own source code also.